Mathlib/RingTheory/Kaehler/JacobiZariski.lean

```diff
@@ -0,0 +1,559 @@
+/-
+Copyright (c) 2024 Andrew Yang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Andrew Yang
+-/
+import Mathlib.RingTheory.Kaehler.CotangentComplex
+import Mathlib.RingTheory.Generators
+import Mathlib.Algebra.Module.SnakeLemma
+
+/-!
+
+# The Jacobi-Zariski exact sequence
+
+Given `R → S → T`, the Jacobi-Zariski exact sequence is
+```
+H¹(L_{T/R}) → H¹(L_{T/S}) → T ⊗[S] Ω[S/R] → Ω[T/R] → Ω[T/S] → 0
+```
+The maps are
+- `Algebra.H1Cotangent.map`
+- `Algebra.H1Cotangent.δ`
+- `KaehlerDifferential.mapBaseChange`
+- `KaehlerDifferential.map`
+and the exactness lemmas are
+- `Algebra.H1Cotangent.exact_map_δ`
+- `Algebra.H1Cotangent.exact_δ_mapBaseChange`
+- `KaehlerDifferential.exact_mapBaseChange_map`
+- `KaehlerDifferential.map_surjective`
+-/
+
+open KaehlerDifferential TensorProduct MvPolynomial
+
+namespace Algebra
+
+universe w' w u v uT
+
+
+variable {R : Type u} {S : Type v} [CommRing R] [CommRing S] [Algebra R S] {P : Generators.{w} R S}
+variable {T : Type uT} [CommRing T] [Algebra R T] [Algebra S T] [IsScalarTower R S T]
+
+attribute [local instance] SMulCommClass.of_commMonoid
+
+attribute [local instance 999999] Zero.toOfNat0 SemilinearMapClass.distribMulActionSemiHomClass
+  SemilinearEquivClass.instSemilinearMapClass TensorProduct.addZeroClass AddZeroClass.toZero
+
+namespace Generators
+
+/--
+Given `R[X] → S` and `S[Y] → T`, this is the lift of an element in `ker(S[Y] → T)`
+to `ker(R[X][Y] → S[Y] → T)`.
+-/
+noncomputable
+def kerCompPreimage (Q : Generators S T) (P : Generators R S) (x : Q.ker) :
+    (Q.comp P).ker := by
+  refine ⟨x.1.sum fun n r ↦ ?_, ?_⟩
+  · refine rename ?_ (P.σ r) * monomial ?_ 1
+    exacts [Sum.inr, n.mapDomain Sum.inl]
```

Comment:
This is intentional to control the elaboration order so that the term has type `(Q.comp P).Ring` and not `MvPolynomial (Q.vars ⊕ P.vars) R`

Reply:
Could you in this case at least add a comment explaining that the spelling is intentional?
Independent of this PR, is there a reason why we need the `abbrev` `Generators.Ring` or is it purely for convenience?

Reply:
We need an algebra instance of `P.Ring` on `S` (if `P : Generators R S`)

Reply:
Agreed, but couldn't we have one on `MvPolynomial P.vars R` instead? This of course has the same potential diamond problem as `P.Ring`, but it might be smoother for `rw` and also less work to maintain the correct `simp` set.

Reply:
`Generator.Ring` is already reducible so I don't think that's a big problem.

The reason why we need to massage the terms here is because `(Q.comp P).vars` is not reducibly defeq to `Q.vars ⊕ P.vars`. I'm thinking if we should make `comp` reducible to fix this.